### PR TITLE
[trainer, worker] feat: grouped mini-batching for multi-output rollouts in main_ppo_sync

### PR DESCRIPTION
**Pull request by guillemgt**
### What does this PR do?

When using multi-output rollouts, mini-batch splitting can break sequences from the same prompt across different PPO mini-batches. Namely, each global batch is split into mini-batches of `actor.ppo_mini_batch_size * rollout.n` sequences, which does not necessarily correspond to the sequences corresponding to `actor.ppo_mini_batch_size` different prompts when using multi-output rollouts.
This also results in an inconsistent number of model updates per training iteration.

This PR fixes this by:
- Adding `make_grouped_iterator` in `tensordict_utils.py` that keeps same-UID sequences together in mini-batches
- Computing `num_mini_batch` from unique prompts instead of `mini_batch_size * rollout.n`.
- Computing `global_batch_size` dynamically via `all_reduce` across DP ranks.

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: [main_ppo_sync](https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+main_ppo_sync)
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

Validated experimentally with multi-output training where each session has a variable number of outputs.

### API and Usage Example

No API changes. The behavior is automatic — `ppo_mini_batch_size` is interpreted as the number of unique prompts per mini-batch, and the grouped iterator keeps all outputs from the same prompt together.

### Design & Code Changes

- **`verl/utils/tensordict_utils.py`**: New `make_grouped_iterator` function that groups sequences by a key (default: `uid`) and splits unique groups into `num_mini_batch` chunks.
- **`verl/workers/engine_workers.py`**: `TrainingWorker.train_batch` uses `make_grouped_iterator` when `num_mini_batch` and `uid` are available. Computes `global_batch_size` dynamically via `all_gather_object` since grouped mini-batches may have variable sizes across DP ranks.
- **`verl/trainer/main_ppo_sync.py`**: `_update_actor` / `_update_critic`: Compute `num_mini_batch` from unique UIDs rather than scaling `mini_batch_size` by `rollout.n`.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs). Not necessary
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: changes are covered by existing tests
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [x] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`. Not applicable

### CODE REVIEW

**Comment by wuxibin89** [comment]
> Trimming the batch when not divisible by dp_size (instead of raising)

https://github.com/verl-project/verl/pull/5969 support dropless padding

**Comment by guillemgt** [comment]
> #5969 support dropless padding

Thanks, I've removed the downsampling changes, but kept the per-sample mini-batch splitting.

File: verl/trainer/main_ppo_sync.py [hunk 1]
```diff
@@ -1009,9 +1009,18 @@ def _balance_batch(self, batch: KVBatchMeta, metrics, logging_prefix="global_seq
...
+            trim_size = len(batch) % dp_size
+            batch = KVBatchMeta(
+                keys=batch.keys[: len(batch) - trim_size],
+                tags=batch.tags[: len(batch) - trim_size],
+                partition_id=batch.partition_id,
+                fields=batch.fields,
+                extra_info=batch.extra_info,
+            )
+            global_seqlen_lst = torch.tensor([tag["seq_len"] for tag in batch.tags], dtype=torch.int64)
+            workload_lst = calculate_workload(global_seqlen_lst)
```
[the remark this answers is not in the document]
Comment (guillemgt) [reply verl/trainer/main_ppo_sync.py]: The downsampling changes have been removed from the PR.

File: verl/workers/engine_workers.py [hunk 2]
```diff
@@ -294,9 +303,20 @@ def train_mini_batch(self, data: TensorDict) -> TensorDict:
...
 
+                local_batch_size = mini_batch_td.batch_size[0]
+                dp_group = self.engine.get_data_parallel_group()
+                dp_world = torch.distributed.get_world_size(dp_group)
+                if dp_world > 1:
+                    sizes = [None] * dp_world
+                    torch.distributed.all_gather_object(sizes, local_batch_size, dp_group)
+                    global_batch_size = sum(sizes)
+                else:
+                    global_batch_size = local_batch_size
```
[the remark this answers is not in the document]
Comment (guillemgt) [reply verl/workers/engine_workers.py:318]: Fixed.

**Comment by guillemgt** [comment]
@wuxibin89 Since #5969 has been merged, could you have another look at this PR?

In the current main, when using multi-output rollouts, the mini-batches are not correctly computed as explained in the (updated) PR description, resulting in rollouts for one prompt being in different mini-batches and each training iteration consisting of a different number of mini-batches.
